File: 2494-sum-of-prefix-scores-of-strings/2494-sum-of-prefix-scores-of-strings.cpp
class Solution {
    class Trie
    {
        public :
        const static int MAXCHAR  = 26;
        Trie* arr[MAXCHAR]{};
        int cnt{}; 
         void add_word(string &str, int idx=0)
        {
            if(idx>=str.size())
                return ;
            int i = str[idx]-'a';
            if(arr[i] == nullptr)
                arr[i] = new Trie();
            arr[i]->cnt++;
            arr[i]->add_word(str, idx+1);
        }
    int is_prefix(string &str, int idx= 0)
        {
            int sum = 0; 
            if(idx>=str.size())
                return 0;
            int i = str[idx]-'a';
            
            if(!arr[i])
                return 0;
            sum += arr[i]->cnt;
            return sum+arr[i]->is_prefix(str, idx+1);
        }
        
    };
   
public:
    vector<int> sumPrefixScores(vector<string>& words) {
        Trie* trie = new Trie();
        for(int i = 0; i< words.size(); i++)
        {
            trie->add_word(words[i]);
        }
        vector<int>result(words.size(),0);
        for(int i = 0; i< words.size(); i++)
        {
            int cnt = 0; 
            
                cnt += trie->is_prefix(words[i]);
            
            result[i]= cnt;
        }return result;
        
    }
};

File: 2494-sum-of-prefix-scores-of-strings/2494-sum-of-prefix-scores-of-strings.cs
class trie
    {
        
         public const int m=26;
        public trie[] child;
        public int cnt=0;
       
        public trie(){
            child=new trie[26];
        }
        
        
        
    };

    




public class Solution {
    public int[] SumPrefixScores(string[] words) {


        trie Trie  = new trie();
        for(int i = 0; i< words.Length; i++)
        {
            insert(Trie, words[i]);
        }
        int[]ans = new int[words.Length];
        for(int i = 0; i< words.Length; i++)
        {
            ans[i]= (word_exist(Trie,words[i],0));

        }
        return ans;

        void insert(trie Trie, string str, int indx=0)
        {
            
            while(indx<str.Length)
            {
                int cur = str[indx]-'a';
                if(Trie.child[cur] == null)
                    Trie.child[cur] = new trie();
                Trie.child[cur].cnt++;
                Trie= Trie.child[cur];
                indx++;
            }
                
        }

        int word_exist(trie Trie,string str, int indx=0)
        {
            if(indx == str.Length)
                return 0;
            
            int x = 0; 
           
            trie f = Trie;
           
            while(indx<str.Length)
            {
                
                x+= f.child[str[indx]-'a'].cnt;
                f=f.child[str[indx]-'a'];
                indx++;
            }
            return x;
        }
        
    }
}










    
    
